package/src/bin/thing-flinger.rs

```diff
@@ -0,0 +1,590 @@
+    if !diff.is_empty() {
+        Err(FlingError::InvalidServers(diff))
+    } else {
+        Ok(())
+    }
+}
+
+// TODO: run in parallel when that option is given
+fn do_exec(
+    config: &Config,
+    cmd: String,
+    servers: Option<BTreeSet<String>>,
+) -> Result<()> {
+    if servers.is_some() {
+        validate_servers(&servers.as_ref().unwrap(), &config.servers)?;
+
+        for name in &servers.unwrap() {
+            let server = &config.servers[name];
+            ssh_exec(&server, &cmd, false)?;
+        }
+    } else {
+        for (_, server) in config.servers.iter() {
+            ssh_exec(&server, &cmd, false)?;
+        }
+    }
+    Ok(())
+}
+
+fn do_sync(config: &Config) -> Result<()> {
+    let server =
+        config.servers.get(&config.builder.server).ok_or_else(|| {
+            FlingError::InvalidServers(vec![config.builder.server.clone()])
+        })?;
+    let src = config.local_source.clone() + "/";
+    let dst = format!(
+        "{}@{}:{}",
+        server.username, server.addr, config.builder.omicron_path
+    );
+    let mut cmd = Command::new("rsync");
+    cmd.arg("-az")
+        .arg("-e")
+        .arg("ssh")
+        .arg("--delete")
+        .arg("--progress")
+        .arg("--exclude")
+        .arg("target/")
+        .arg("--exclude")
+        .arg("out/")
+        .arg("--exclude")
+        .arg("cockroachdb/")
+        .arg("--exclude")
+        .arg("clickhouse/")
+        .arg("--exclude")
+        .arg("*.swp")
+        .arg("--out-format")
+        .arg("File changed: %o %t %f")
+        .arg(&src)
+        .arg(&dst);
+    let status =
+        cmd.status().context(format!("Failed to run command: ({:?})", cmd))?;
+    if !status.success() {
+        return Err(FlingError::FailedSync { src, dst }.into());
+    }
+
+    Ok(())
+}
+
+// Build omicron-package and omicron-sled-agent on the builder
+//
+// We need to build omicron-sled-agent for overlay file generation
+fn do_build_minimal(config: &Config) -> Result<()> {
+    let server = &config.servers[&config.builder.server];
+    let cmd = format!(
+        "cd {} && git checkout {} && cargo build -p {} -p {}",
+        config.builder.omicron_path,
+        config.builder.git_treeish,
+        "omicron-package",
+        "omicron-sled-agent"
+    );
+    ssh_exec(&server, &cmd, false)
+}
+
+fn do_package(
+    config: &Config,
+    artifact_dir: PathBuf,
+    release: bool,
+) -> Result<()> {
+    let server = &config.servers[&config.builder.server];
+    let mut cmd = String::new();
+    let mut release_flag = "";
+    if release {
+        release_flag = "--release";
+    }
+    let cmd_path = "./target/debug/omicron-package";
+    let artifact_dir = artifact_dir
+        .to_str()
+        .ok_or_else(|| FlingError::BadString("artifact_dir".to_string()))?;
+
+    // We use a bash login shell to get a proper environment, so we have a path to
+    // postgres, and $DEP_PQ_LIBDIRS is filled in. This is required for building
```

Comment:
Does this rely on some part of postgres being on the remote shell's `PATH`? I might include a note to that effect in the README or command help.

Reply:
Good Point! This is because the builder must have all the proper software installed per the Omicron README. There's a note in the README pre-requisites to this effect, but I made it stronger just now.